openedx/core/lib/x_forwarded_for/middleware.py

```diff
@@ -108,3 +87,26 @@ def process_request(self, request):
             request.META['REMOTE_ADDR'] = legacy_ip.get_legacy_ip(request)
         else:
             request.META['REMOTE_ADDR'] = ip.get_safest_client_ip(request)
+
+        try:
+            # Give some observability into IP chain length and composition. Useful
+            # for monitoring in case of unexpected network config changes, etc.
+            ip_chain = ip.get_raw_ip_chain(request)
+
+            # .. custom_attribute_name: ip_chain.raw
+            # .. custom_attribute_description: The actual contents of the raw IP chain. Could
+            #      be used to correlate authenticated and unauthenticated requests for the same
+            #      user.
+            set_custom_attribute('ip_chain.raw', ', '.join(ip_chain))
+            set_custom_attribute('ip_chain.count', len(ip_chain))
+            set_custom_attribute('ip_chain.types', '-'.join(_ip_type(s) for s in ip_chain))
+
+            set_custom_attribute('ip_chain.use_legacy', legacy_ip.USE_LEGACY_IP.is_enabled())
+
+            external_chain = ip.get_all_client_ips(request)
+            set_custom_attribute('ip_chain.external.count', len(external_chain))
+            set_custom_attribute('ip_chain.external.types', '-'.join(_ip_type(s) for s in external_chain))
+            
+            set_custom_attribute('ip_chain.remote_addr', request.META['REMOTE_ADDR'])
```

Comment:
This line is the only new line. All other lines are a move.

**Why the move?** Opted to avoid re-calculating or repeating logic of defining `request.META['REMOTE_ADDR']`, and didn't immediately see any drawback to setting the custom attributes after the set of `request.META['REMOTE_ADDR']`.

Reply:
I think we should prefer to record the output of `get_safest_client_ip`, which is a more well-defined value. `REMOTE_ADDR` is an artifact of this middleware and depends on the value of a Waffle switch. It could be useful to know what value is in effect, but we're trying to get rid of legacy_ip. (Actually, we should be able to do so by now? I'll get that prioritized.)

Some background: `REMOTE_ADDR` is intended to be the IP address on the other end of the immediate TCP connection, and is usually not a useful value. This middleware had been overriding it to the leftmost IP in the chain, which is spoofable. This makes it impossible to reconstruct the _actual_ IP chain, and oversimplifies the situation for relying code, which is... not great.

I think what I'd prefer is to lift the call to `get_safest_client_ip` out of the `if` block so it can be used in both the second branch and in a custom attribute. The custom attribute should probably be called something like `ip_chain.safest_client_ip` -- it's the most conservative guess.

Reply:
([inform] `USE_LEGACY_IP` is disabled in our current production configuration.)

Reply:
Makes sense! Refactored in https://github.com/openedx/edx-platform/pull/33720/commits/29d549259cd63821a8b0e293a3a4ccacc05fac78.